src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,519 @@
+        /// <summary>
+        /// Executes the <paramref name="entry"/>.
+        /// </summary>
+        /// <param name="entry">Transaction entry to be executed inside the <see cref="VirtualMachine"/>.</param>
+        /// <param name="stateUpdate"><see cref="Delta"/> to be used for execution environment construction</param>
+        /// <param name="txTracer">Tracer to extract the execution steps for debugging or analytics.</param>
+        /// <param name="readOnly">Defines whether the state should be reverted after the execution.</param>
+        /// <exception cref="TransactionCollisionException">Thrown when deployment address already has some code.</exception>
+        /// <exception cref="OutOfGasException">Thrown when not enough gas is available for deposit.</exception>
+        private void Execute(ContractEntry entry, StateUpdate stateUpdate, ITxTracer txTracer, bool readOnly)
+        {
+            IReleaseSpec spec = _specProvider.GetSpec(stateUpdate.Number);
+
+            (Address sender, Address recipient) = ExtractSenderAndRecipient(entry);
+            bool isPrecompile = recipient.IsPrecompiled(spec);
+            ExecutionEnvironment env = PrepareEnv(entry, sender, recipient, stateUpdate, isPrecompile);
+
+            ulong gasLimit = entry.GasLimit;
+            ulong intrinsicGas = CalculateIntrinsicGas(entry, spec);
+
+            if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Executing entry {entry}", entry);
+            if (!ValidateSender(entry, env, txTracer)) return;
+            if (!ValidateIntrinsicGas(entry, env, intrinsicGas, txTracer)) return;
+            if (!ValidateDeltaGasLimit(entry, env, txTracer)) return;
+
+            if (!_stateProvider.AccountExists(env.Sender))
+            {
+                if (env.GasPrice == UInt256.Zero) _stateProvider.CreateAccount(env.Sender, UInt256.Zero);
+            }
+
+            if (!ValidateSenderBalance(entry, env, intrinsicGas, txTracer)) return;
+            if (!ValidateNonce(entry, env, txTracer)) return;
+
+            InitEntryExecution(env, gasLimit, spec, txTracer);
+
+            // we prepare two fields to track the amount of gas spent / left
+            ulong unspentGas = gasLimit - intrinsicGas;
+            ulong spentGas = gasLimit;
+
+            // the snapshots are needed to revert the subroutine state changes in case of an VM exception
+            int stateSnapshot = _stateProvider.TakeSnapshot();
+            int storageSnapshot = _storageProvider.TakeSnapshot();
+
+            // we subtract value from sender
+            // it will be added to recipient at the later stage (inside the VM)
+            _stateProvider.SubtractFromBalance(sender, env.Value, spec);
+
+            // we fail unless we succeed
+            byte statusCode = StatusCode.Failure;
+            TransactionSubstate substate = null;
+
+            try
+            {
+                if (entry.IsValidDeploymentEntry) PrepareContractAccount(env.CodeSource);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)